kubernetes/cache.go

```diff
@@ -0,0 +1,325 @@
+	"k8s.io/api/apps/v1beta2"
+	batch_v1 "k8s.io/api/batch/v1"
+	batch_v1beta1 "k8s.io/api/batch/v1beta1"
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	kube "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type (
+	// Inspired/reused from istio code:
+	// https://github.com/istio/istio/blob/master/mixer/adapter/kubernetesenv/cache.go
+	cacheController interface {
+		// Control Cache
+		Run(<-chan struct{})
+		HasSynced() bool
+		StopControlChannel()
+
+		// Business methods
+		GetCronJobs(namespace string) ([]batch_v1beta1.CronJob, bool)
+		GetDeployment(namespace string, name string) (*v1beta1.Deployment, bool)
+		GetDeployments(namespace string) ([]v1beta1.Deployment, bool)
+		GetEndpoints(namespace, name string) (*v1.Endpoints, bool)
+		GetJobs(namespace string) ([]batch_v1.Job, bool)
+		GetPods(namespace string) ([]v1.Pod, bool)
+		GetReplicationControllers(namespace string) ([]v1.ReplicationController, bool)
+		GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, bool)
+		GetService(namespace string, name string) (*v1.Service, bool)
+		GetServices(namespace string) ([]v1.Service, bool)
+		GetStatefulSet(namespace string, name string) (*v1beta2.StatefulSet, bool)
+		GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, bool)
+	}
+
+	controllerImpl struct {
+		stopChan     chan struct{}
+		syncCount    int
+		maxSyncCount int
+		controllers  map[string]cache.SharedIndexInformer
+	}
+)
+
+func newCacheController(clientset kube.Interface, refreshDuration time.Duration, stopChan chan struct{}) cacheController {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, refreshDuration)
+	controllers := make(map[string]cache.SharedIndexInformer)
+
+	controllers["Pod"] = sharedInformers.Core().V1().Pods().Informer()
+	controllers["ReplicationController"] = sharedInformers.Core().V1().ReplicationControllers().Informer()
+	controllers["Deployment"] = sharedInformers.Apps().V1beta1().Deployments().Informer()
+	controllers["ReplicaSet"] = sharedInformers.Apps().V1beta2().ReplicaSets().Informer()
+	controllers["StatefulSet"] = sharedInformers.Apps().V1beta2().StatefulSets().Informer()
+	controllers["Job"] = sharedInformers.Batch().V1().Jobs().Informer()
+	controllers["CronJob"] = sharedInformers.Batch().V1beta1().CronJobs().Informer()
+	controllers["Service"] = sharedInformers.Core().V1().Services().Informer()
+	controllers["Endpoints"] = sharedInformers.Core().V1().Endpoints().Informer()
+
+	return &controllerImpl{
+		stopChan:     stopChan,
+		controllers:  controllers,
+		syncCount:    0,
+		maxSyncCount: 20, // Move this to config ? or this constant is good enough ?
+	}
+}
+
+func (c *controllerImpl) Run(stop <-chan struct{}) {
+	for _, cn := range c.controllers {
+		go cn.Run(stop)
+	}
+	<-stop
+}
+
+func (c *controllerImpl) HasSynced() bool {
+	if c.syncCount > c.maxSyncCount {
+		log.Errorf("Max attempts reached syncing cache. Is there connection with the K8S backend ?")
+		c.StopControlChannel()
```

Comment:
what happens here exactly? Are we entering in an unrecoverable state by stopping control channel?

Reply:
On initialization, I see usual to have an active wait until cache is populated:
```
cache.WaitForCacheSync(client.stopCache, client.k8sCache.HasSynced)
```
(I saw this on tests and also in usage of the cache)

But it happens that there is a cornercase when the backend is always down, so, to resolve this properly I forced to fail.

In the preliminar testing I did, it looked good.
But with your comments and also other ideas I thought I guess cache could hide better the stopchannel and provide a simple way for this; I'm refactoring that part, I'll update the PR later.